Add health check support for Keycloak container

paulomorgado opened the pull request:
## Description

This PR introduces new constants for health checks, including an environment variable (`KC_HEALTH_ENABLED`), a health check container port (9000), and a health endpoint name (`health`). The `AddKeycloak` method is updated to incorporate a health check endpoint and set the health check environment variable to "true" for the Keycloak resource.

Fixes #6113

/cc @davidfowl @DamianEdwards 

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [x] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [x] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Is this introducing a breaking change?
      - [ ] Yes
        - Link to aspire-docs issue (please use this [`breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)):
      - [ ] No
        - Link to aspire-docs issue (please use this [`doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)):
  - [ ] No

paulomorgado committed:
Add health check support for Keycloak container

This commit introduces new constants for health checks, including
an environment variable (`KC_HEALTH_ENABLED`), a health check
container port (9000), and a health endpoint name (`health`).
The `AddKeycloak` method is updated to incorporate a health
check endpoint and set the health check environment variable
to "true" for the Keycloak resource.

paulomorgado committed:
Refactor Keycloak tests for endpoint annotations and manifest

Updated `AddKeycloakWithDefaultsAddsAnnotationMetadata` to filter `EndpointAnnotation` by name for "http" and "health", adding assertions for the "health" endpoint. Modified `VerifyManifest` to use a specific image reference and added the `KC_HEALTH_ENABLED` environment variable, along with updating the `bindings` section to include the "health" binding.

davidfowl commented on tests/Aspire.Hosting.Keycloak.Tests/KeycloakResourceBuilderTests.cs:
```diff
@@ -135,21 +144,28 @@ public async Task VerifyManifest()
             {
               "type": "container.v0",
-              "image": "{{KeycloakContainerImageTags.Registry}}/{{KeycloakContainerImageTags.Image}}:{{KeycloakContainerImageTags.Tag}}",
+              "image": "quay.io/keycloak/keycloak:26.0",
```
Why did you change this?

paulomorgado replied:
OOPS! Reverted!

davidfowl commented:
@DamianEdwards can you review?

paulomorgado committed:
Revert change made by mistake.

DamianEdwards commented on src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs:
```diff
@@ -14,7 +14,10 @@ public static class KeycloakResourceBuilderExtensions
     private const string AdminPasswordEnvVarName = "KEYCLOAK_ADMIN_PASSWORD";
+    private const string HealthCheckEnvVarName = "KC_HEALTH_ENABLED";
     private const int DefaultContainerPort = 8080;
+    private const int HealthCheckContainerPort = 9000;
```
Seems Keycloak refers to this endpoint as the "management interface": https://www.keycloak.org/server/management-interface

Let's change the names here accordingly:

```suggestion
    private const string ManagementInterfaceContainerPort = 9000;
```

DamianEdwards commented on src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs:
```diff
@@ -14,7 +14,10 @@ public static class KeycloakResourceBuilderExtensions
+    private const string HealthCheckEnvVarName = "KC_HEALTH_ENABLED";
     private const int DefaultContainerPort = 8080;
+    private const int HealthCheckContainerPort = 9000;
+    private const string HealthEndpointName = "health";
```
Per above:

```suggestion
    private const string ManagementEndpointName = "management";
```

DamianEdwards commented on src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs:
```diff
@@ -59,10 +62,13 @@ public static IResourceBuilder<KeycloakResource> AddKeycloak(
             .WithImageRegistry(KeycloakContainerImageTags.Registry)
             .WithImageTag(KeycloakContainerImageTags.Tag)
             .WithHttpEndpoint(port: port, targetPort: DefaultContainerPort)
+            .WithHttpEndpoint(targetPort: HealthCheckContainerPort, name: HealthEndpointName)
```
```suggestion
            .WithHttpEndpoint(targetPort: ManagementInterfaceContainerPort, name: ManagementEndpointName)
```

DamianEdwards commented on src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs:
```diff
@@ -59,10 +62,13 @@ public static IResourceBuilder<KeycloakResource> AddKeycloak(
             .WithImageTag(KeycloakContainerImageTags.Tag)
             .WithHttpEndpoint(port: port, targetPort: DefaultContainerPort)
+            .WithHttpEndpoint(targetPort: HealthCheckContainerPort, name: HealthEndpointName)
+            .WithHttpHealthCheck(endpointName: HealthEndpointName, path: "/health/ready")
```
```suggestion
            .WithHttpHealthCheck(endpointName: ManagementEndpointName, path: "/health/ready")
```

DamianEdwards commented on src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs:
```diff
@@ -14,7 +14,10 @@ public static class KeycloakResourceBuilderExtensions
 {
     private const string AdminEnvVarName = "KEYCLOAK_ADMIN";
     private const string AdminPasswordEnvVarName = "KEYCLOAK_ADMIN_PASSWORD";
+    private const string HealthCheckEnvVarName = "KC_HEALTH_ENABLED";
```
Comment with a link to the relevant docs? e.g.

```csharp
// As per https://www.keycloak.org/observability/health
```

paulomorgado replied:
Done!

DamianEdwards commented on src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs:
```diff
@@ -14,7 +14,10 @@ public static class KeycloakResourceBuilderExtensions
     private const string AdminPasswordEnvVarName = "KEYCLOAK_ADMIN_PASSWORD";
+    private const string HealthCheckEnvVarName = "KC_HEALTH_ENABLED";
     private const int DefaultContainerPort = 8080;
+    private const int HealthCheckContainerPort = 9000;
```
Comment with a link to the relevant docs? e.g.

```csharp
As per https://www.keycloak.org/server/management-interface
```

paulomorgado replied:
Done!

DamianEdwards commented on src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs:
```diff
@@ -59,10 +62,13 @@ public static IResourceBuilder<KeycloakResource> AddKeycloak(
             .WithImageRegistry(KeycloakContainerImageTags.Registry)
             .WithImageTag(KeycloakContainerImageTags.Tag)
             .WithHttpEndpoint(port: port, targetPort: DefaultContainerPort)
+            .WithHttpEndpoint(targetPort: HealthCheckContainerPort, name: HealthEndpointName)
```
Consider adding the management interface port as an optional parameter too (like `port`) so it can be easily changed by the caller. If it's set then we'd need to set the `KC_HTTP_MANAGEMENT_PORT` environment variable too.

paulomorgado replied:
Wouldn't that mean that we should also support `KC_HTTP_PORT`?

But maybe we could add a `managementPort` parameter with the same semantics of the `port` parameter, without messing with target ports.

Would it be feasible to determine the target ports based on the existence or not of `KC_HTTP_PORT` and `KC_HTTP_MANAGEMENT_PORT`?

DamianEdwards replied:
> But maybe we could add a managementPort parameter with the same semantics of the port parameter, without messing with target ports.

Yeah this is what I think I meant 😄 Just let the user specify a different management port which we'll assign to the proxy port. Target port will stay the same always (unless user manually edits the added the endpoint, which of course they can still do).

paulomorgado replied:
Added with tests!

paulomorgado committed:
Update src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs

Co-authored-by: Damian Edwards <damian@damianedwards.com>

paulomorgado committed:
Update src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs

Co-authored-by: Damian Edwards <damian@damianedwards.com>

paulomorgado committed:
Update src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs

Co-authored-by: Damian Edwards <damian@damianedwards.com>

paulomorgado committed:
Update src/Aspire.Hosting.Keycloak/KeycloakResourceBuilderExtensions.cs

Co-authored-by: Damian Edwards <damian@damianedwards.com>

paulomorgado committed:
Refactor Keycloak resource constants and tests

- Retained `HealthCheckEnvVarName` constant; changed type of `ManagementInterfaceContainerPort` to `int` for clarity.
- Introduced `defaultEndpointName` and `managementEndpointName` variables in tests to replace hardcoded strings for improved readability.
- Updated JSON structure in `VerifyManifest` to change key from "health" to "management" to align with new naming conventions.

davidfowl committed:
Merge branch 'main' into Keycloak-HealthChecks
